onebusaway-android/src/main/java/org/onebusaway/android/travelbehavior/TravelBehaviorManager.java

```diff
@@ -0,0 +1,340 @@
+                            optOutUser();
+                        })
+                .create().show();
+    }
+
+    private String getHtmlConsentDocument() {
+        InputStream inputStream = mApplicationContext.getResources().
+                openRawResource(R.raw.travel_behavior_informed_consent);
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+
+        byte buf[] = new byte[1024];
+        int len;
+        try {
+            while ((len = inputStream.read(buf)) != -1) {
+                outputStream.write(buf, 0, len);
+            }
+            outputStream.close();
+            inputStream.close();
+        } catch (IOException e) {
+
+        }
+        return outputStream.toString();
+    }
+
+    private void showEmailDialog() {
+        showEmailDialog(null);
+    }
+
+    private void showEmailDialog(String email) {
+        EditText emailEditText = new EditText(mActivityContext);
+        emailEditText.setHint(R.string.travel_behavior_email_hint);
+        emailEditText.setInputType(InputType.TYPE_CLASS_TEXT |
+                InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS);
+
+        if (email != null) {
+            emailEditText.setText(email);
+        }
+
+        new AlertDialog.Builder(mActivityContext)
+                .setTitle(R.string.travel_behavior_opt_in_title)
+                .setMessage(R.string.travel_behavior_email_message)
+                .setIcon(createIcon())
+                .setCancelable(false)
+                .setView(emailEditText)
+                .setPositiveButton(R.string.travel_behavior_dialog_email_save,
+                        (dialog, which) -> {
+                            String currentEmail = emailEditText.getText().toString();
+                            if (!TextUtils.isEmpty(currentEmail) &&
+                                    Patterns.EMAIL_ADDRESS.matcher(currentEmail).matches()) {
+                                registerUser(currentEmail);
+                                checkPermissions();
+                            } else {
+                                Toast.makeText(mApplicationContext, R.string.travel_behavior_email_invalid,
+                                        Toast.LENGTH_LONG).show();
+                                // Android automatically dismisses the dialog.
+                                // Show the dialog again if the email is invalid
+                                showEmailDialog(currentEmail);
+                            }
+                        })
+                .create().show();
+    }
+
+    private void checkPermissions() {
+        if (!PermissionUtils.hasGrantedPermissions(mApplicationContext, TravelBehaviorConstants.PERMISSIONS)) {
+            Activity homeActivity = (Activity) mActivityContext;
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                homeActivity.requestPermissions(TravelBehaviorConstants.PERMISSIONS, 1);
+            }
+        }
+    }
+
```

Comment:
@barbeau I implemented the permission check on registration.  In the current implementation, we check the permission and do nothing if the user does not allow the permission. What do you think?

Reply:
Great!  I think that's fine.